rbd: fix mapOptions passing with rbd-nbd mounter

pkalever opened the pull request:
<!-- Please take a look at our [Contributing](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#Code-contribution-workflow)
documentation before submitting a Pull Request!
Thank you for contributing to ceph-csi! -->

# Describe what this PR does #
* rbd: fix mapOptions passing with rbd-nbd mounter
This was a regression introduced by:
https://github.com/ceph/ceph-csi/pull/2556

* e2e: adding io-timeout for lower kernel versions
This got removed unintentionally with
https://github.com/ceph/ceph-csi/pull/2628

* e2e: validate encrypted image mount inside the nodeplugin
currently, the mount type validation of the encrypted volume is done in the application, we should rather validate this inside the nodeplugin pod.

Signed-off-by: Prasanna Kumar Kalever <prasanna.kalever@redhat.com>


Fixes: #2610.
Signed-off-by: Prasanna Kumar Kalever \<prasanna.kalever@redhat.com\>


---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

- `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)
- `/retest all`: run this in case the CentOS CI failed to start/report any test
  progress or results

</details>

Commit:
e2e: adding io-timeout for lower kernel versions

This got removed unintentionally with
https://github.com/ceph/ceph-csi/pull/2628

Signed-off-by: Prasanna Kumar Kalever <prasanna.kalever@redhat.com>

Commit:
rbd: fix mapOptions passing with rbd-nbd mounter

This was a regression introduced by:
https://github.com/ceph/ceph-csi/pull/2556

Fixes: #2610
Signed-off-by: Prasanna Kumar Kalever <prasanna.kalever@redhat.com>

pkalever commented:
All the tests are hitting a different common issue:

https://jenkins-ceph-csi.apps.ocp.ci.centos.org/job/mini-e2e-helm_k8s-1.20/3078/display/redirect
https://jenkins-ceph-csi.apps.ocp.ci.centos.org/job/mini-e2e-helm_k8s-1.21/1737/display/redirect
https://jenkins-ceph-csi.apps.ocp.ci.centos.org/job/mini-e2e-helm_k8s-1.22/1027/display/redirect

https://jenkins-ceph-csi.apps.ocp.ci.centos.org/job/mini-e2e_k8s-1.20/3096/display/redirect
https://jenkins-ceph-csi.apps.ocp.ci.centos.org/job/mini-e2e_k8s-1.21/1753/display/redirect
https://jenkins-ceph-csi.apps.ocp.ci.centos.org/job/mini-e2e_k8s-1.22/1004/display/redirect

```
[1mSTEP[0m: create a PVC and bind it to an app using rbd-nbd mounter with encryption
Nov 14 09:47:49.419: INFO: waiting for kubectl (delete -f args []) to finish
Nov 14 09:47:49.419: INFO: Running '/usr/bin/kubectl --server=https://192.168.39.54:8443 --kubeconfig=/root/.kube/config --namespace=cephcsi-e2e-2ea9e9c277d7 delete -f -'
Nov 14 09:47:49.491: INFO: stderr: "warning: deleting cluster-scoped resources, not scoped to the provided namespace\n"
Nov 14 09:47:49.491: INFO: stdout: "storageclass.storage.k8s.io \"csi-rbd-sc\" deleted\n"
Nov 14 09:47:49.496: INFO: ExecWithOptions {Command:[/bin/sh -c ceph fsid] Namespace:rook-ceph PodName:rook-ceph-tools-7467d8bf8-wjvdf ContainerName:rook-ceph-tools Stdin:<nil> CaptureStdout:true CaptureStderr:true PreserveWhitespace:true Quiet:false}
Nov 14 09:47:49.496: INFO: >>> kubeConfig: /root/.kube/config
Nov 14 09:47:50.414: INFO: Waiting up to &PersistentVolumeClaim{ObjectMeta:{rbd-pvc  rbd-1318    0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] []  []},Spec:PersistentVolumeClaimSpec{AccessModes:[ReadWriteOnce],Resources:ResourceRequirements{Limits:ResourceList{},Requests:ResourceList{storage: {{1073741824 0} {<nil>} 1Gi BinarySI},},},VolumeName:,Selector:nil,StorageClassName:*csi-rbd-sc,VolumeMode:nil,DataSource:nil,DataSourceRef:nil,},Status:PersistentVolumeClaimStatus{Phase:,AccessModes:[],Capacity:ResourceList{},Conditions:[]PersistentVolumeClaimCondition{},},} to be in Bound state
Nov 14 09:47:50.414: INFO: waiting for PVC rbd-pvc (0 seconds elapsed)
Nov 14 09:47:52.417: INFO: waiting for PVC rbd-pvc (2 seconds elapsed)
Nov 14 09:47:52.421: INFO: Waiting for PV pvc-427711d5-35cc-4e42-a0a1-b2097903c783 to bind to PVC rbd-pvc
Nov 14 09:47:52.421: INFO: Waiting up to timeout=10m0s for PersistentVolumeClaims [rbd-pvc] to have phase Bound
Nov 14 09:47:52.423: INFO: PersistentVolumeClaim rbd-pvc found and phase=Bound (1.923219ms)
Nov 14 09:47:52.423: INFO: Waiting up to 10m0s for PersistentVolume pvc-427711d5-35cc-4e42-a0a1-b2097903c783 to have phase Bound
Nov 14 09:47:52.425: INFO: PersistentVolume pvc-427711d5-35cc-4e42-a0a1-b2097903c783 found and phase=Bound (2.076567ms)
Nov 14 09:47:52.432: INFO: Waiting up to csi-rbd-demo-pod to be in Running state
Nov 14 09:48:20.444: INFO: ExecWithOptions {Command:[/bin/sh -c rbd image-meta get replicapool/csi-vol-ee3d6f50-452f-11ec-8fc8-9acf4e9aad26 rbd.csi.ceph.com/encrypted] Namespace:rook-ceph PodName:rook-ceph-tools-7467d8bf8-wjvdf ContainerName:rook-ceph-tools Stdin:<nil> CaptureStdout:true CaptureStderr:true PreserveWhitespace:true Quiet:false}
Nov 14 09:48:20.444: INFO: >>> kubeConfig: /root/.kube/config
Nov 14 09:48:20.839: INFO: ExecWithOptions {Command:[/bin/sh -c lsblk -o TYPE,MOUNTPOINT | grep '/var/lib/www/html' | awk '{print $1}'] Namespace:rbd-1318 PodName:csi-rbd-demo-pod ContainerName:web-server Stdin:<nil> CaptureStdout:true CaptureStderr:true PreserveWhitespace:true Quiet:false}
Nov 14 09:48:20.839: INFO: >>> kubeConfig: /root/.kube/config
Nov 14 09:48:20.932: FAIL: failed to validate encrypted pvc with error  not equal to crypt
```

nodeplugin logs:
```
I1114 09:48:09.587996   36302 cephcmds.go:62] ID: 7 Req-ID: 0001-0024-cfb024e1-4eee-4917-b9ec-8acb0f4afcb8-0000000000000001-ee3d6f50-452f-11ec-8fc8-9acf4e9aad26 command succeeded: rbd [--id cephcsi-rbd-node -m rook-ceph-mon-a.rook-ceph.svc.cluster.local:6789 --keyfile=***stripped*** --log-file /var/log/ceph/rbd-nbd-0001-0024-cfb024e1-4eee-4917-b9ec-8acb0f4afcb8-0000000000000001-ee3d6f50-452f-11ec-8fc8-9acf4e9aad26.log map replicapool/csi-vol-ee3d6f50-452f-11ec-8fc8-9acf4e9aad26 --device-type nbd --options try-netlink --options reattach-timeout=300 --options debug-rbd=20,io-timeout=330]
I1114 09:48:09.588031   36302 nodeserver.go:401] ID: 7 Req-ID: 0001-0024-cfb024e1-4eee-4917-b9ec-8acb0f4afcb8-0000000000000001-ee3d6f50-452f-11ec-8fc8-9acf4e9aad26 rbd image: 0001-0024-cfb024e1-4eee-4917-b9ec-8acb0f4afcb8-0000000000000001-ee3d6f50-452f-11ec-8fc8-9acf4e9aad26/replicapool was successfully mapped at /dev/nbd0
I1114 09:48:09.604305   36302 encryption.go:80] ID: 7 Req-ID: 0001-0024-cfb024e1-4eee-4917-b9ec-8acb0f4afcb8-0000000000000001-ee3d6f50-452f-11ec-8fc8-9acf4e9aad26 image replicapool/csi-vol-ee3d6f50-452f-11ec-8fc8-9acf4e9aad26 encrypted state metadata reports "encryptionPrepared"
I1114 09:48:09.604346   36302 mount_linux.go:463] Attempting to determine if disk "/dev/nbd0" is formatted using blkid with args: ([-p -s TYPE -s PTTYPE -o export /dev/nbd0])
I1114 09:48:09.606317   36302 mount_linux.go:466] Output: ""
I1114 09:48:09.606363   36302 crypto.go:199] ID: 7 Req-ID: 0001-0024-cfb024e1-4eee-4917-b9ec-8acb0f4afcb8-0000000000000001-ee3d6f50-452f-11ec-8fc8-9acf4e9aad26 Encrypting device /dev/nbd0 with LUKS
I1114 09:48:15.444350   36302 crypto.go:254] ID: 7 Req-ID: 0001-0024-cfb024e1-4eee-4917-b9ec-8acb0f4afcb8-0000000000000001-ee3d6f50-452f-11ec-8fc8-9acf4e9aad26 device /dev/mapper/luks-rbd-0001-0024-cfb024e1-4eee-4917-b9ec-8acb0f4afcb8-0000000000000001-ee3d6f50-452f-11ec-8fc8-9acf4e9aad26 is not an active LUKS device: an error (exit status 4) occurred while running cryptsetup args: [status luks-rbd-0001-0024-cfb024e1-4eee-4917-b9ec-8acb0f4afcb8-0000000000000001-ee3d6f50-452f-11ec-8fc8-9acf4e9aad26]
I1114 09:48:15.444391   36302 crypto.go:209] ID: 7 Req-ID: 0001-0024-cfb024e1-4eee-4917-b9ec-8acb0f4afcb8-0000000000000001-ee3d6f50-452f-11ec-8fc8-9acf4e9aad26 Opening device /dev/nbd0 with LUKS on luks-rbd-0001-0024-cfb024e1-4eee-4917-b9ec-8acb0f4afcb8-0000000000000001-ee3d6f50-452f-11ec-8fc8-9acf4e9aad26
I1114 09:48:17.189119   36302 mount_linux.go:463] Attempting to determine if disk "/dev/mapper/luks-rbd-0001-0024-cfb024e1-4eee-4917-b9ec-8acb0f4afcb8-0000000000000001-ee3d6f50-452f-11ec-8fc8-9acf4e9aad26" is formatted using blkid with args: ([-p -s TYPE -s PTTYPE -o export /dev/mapper/luks-rbd-0001-0024-cfb024e1-4eee-4917-b9ec-8acb0f4afcb8-0000000000000001-ee3d6f50-452f-11ec-8fc8-9acf4e9aad26])
I1114 09:48:17.217169   36302 mount_linux.go:466] Output: ""
I1114 09:48:17.319419   36302 mount_linux.go:463] Attempting to determine if disk "/dev/mapper/luks-rbd-0001-0024-cfb024e1-4eee-4917-b9ec-8acb0f4afcb8-0000000000000001-ee3d6f50-452f-11ec-8fc8-9acf4e9aad26" is formatted using blkid with args: ([-p -s TYPE -s PTTYPE -o export /dev/mapper/luks-rbd-0001-0024-cfb024e1-4eee-4917-b9ec-8acb0f4afcb8-0000000000000001-ee3d6f50-452f-11ec-8fc8-9acf4e9aad26])
I1114 09:48:17.352173   36302 mount_linux.go:466] Output: "DEVNAME=/dev/mapper/luks-rbd-0001-0024-cfb024e1-4eee-4917-b9ec-8acb0f4afcb8-0000000000000001-ee3d6f50-452f-11ec-8fc8-9acf4e9aad26\nTYPE=ext4\n"
I1114 09:48:17.352223   36302 mount_linux.go:357] Checking for issues with fsck on disk: /dev/mapper/luks-rbd-0001-0024-cfb024e1-4eee-4917-b9ec-8acb0f4afcb8-0000000000000001-ee3d6f50-452f-11ec-8fc8-9acf4e9aad26
I1114 09:48:17.418496   36302 mount_linux.go:453] Attempting to mount disk /dev/mapper/luks-rbd-0001-0024-cfb024e1-4eee-4917-b9ec-8acb0f4afcb8-0000000000000001-ee3d6f50-452f-11ec-8fc8-9acf4e9aad26 in ext4 format at /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-427711d5-35cc-4e42-a0a1-b2097903c783/globalmount/0001-0024-cfb024e1-4eee-4917-b9ec-8acb0f4afcb8-0000000000000001-ee3d6f50-452f-11ec-8fc8-9acf4e9aad26
I1114 09:48:17.418561   36302 mount_linux.go:180] Mounting cmd (mount) with arguments (-t ext4 -o _netdev,discard,defaults /dev/mapper/luks-rbd-0001-0024-cfb024e1-4eee-4917-b9ec-8acb0f4afcb8-0000000000000001-ee3d6f50-452f-11ec-8fc8-9acf4e9aad26 /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-427711d5-35cc-4e42-a0a1-b2097903c783/globalmount/0001-0024-cfb024e1-4eee-4917-b9ec-8acb0f4afcb8-0000000000000001-ee3d6f50-452f-11ec-8fc8-9acf4e9aad26)
I1114 09:48:17.430089   36302 nodeserver.go:361] ID: 7 Req-ID: 0001-0024-cfb024e1-4eee-4917-b9ec-8acb0f4afcb8-0000000000000001-ee3d6f50-452f-11ec-8fc8-9acf4e9aad26 rbd: successfully mounted volume 0001-0024-cfb024e1-4eee-4917-b9ec-8acb0f4afcb8-0000000000000001-ee3d6f50-452f-11ec-8fc8-9acf4e9aad26 to stagingTargetPath /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-427711d5-35cc-4e42-a0a1-b2097903c783/globalmount/0001-0024-cfb024e1-4eee-4917-b9ec-8acb0f4afcb8-0000000000000001-ee3d6f50-452f-11ec-8fc8-9acf4e9aad26
```


Which is different from what we have seen before!

pkalever commented:
@ceph/ceph-csi-contributors I have reordered the e2e tests in a way that krbd encryption tests runs first, just to prove that the bug is not nbd specific. See latest CI results. 

I guess we introduced yet another regression in encryption. Anyone know about it?

Madhu-1 commented:
> I have reordered the e2e tests in a way that krbd encryption tests runs first, just to prove that the bug is not nbd specific. See latest CI results.
> 
> I guess we introduced yet another regression in encryption. Anyone know about it?

but why the E2E is failing only for nbd? can we open a PR to disable nbd and test encryption with krbd to verify the regression?

pkalever commented:
> > I have reordered the e2e tests in a way that krbd encryption tests runs first, just to prove that the bug is not nbd specific. See latest CI results.
> > I guess we introduced yet another regression in encryption. Anyone know about it?
> 
> but why the E2E is failing only for nbd? can we open a PR to disable nbd and test encryption with krbd to verify the regression?

@Madhu-1 sorry, I didn't get you.
Have you missed seeing that the latest e2e tests are failing at test `create a PVC and bind it to an app with encrypted RBD volume` which is not NBD specific and uses krbd mounter only!

```
[1mSTEP[0m: create a PVC and bind it to an app with encrypted RBD volume
Nov 14 11:37:04.267: INFO: waiting for kubectl (delete -f args []) to finish
Nov 14 11:37:04.267: INFO: Running '/usr/bin/kubectl --server=https://192.168.39.7:8443 --kubeconfig=/root/.kube/config --namespace=cephcsi-e2e-8b66e001383f delete -f -'
Nov 14 11:37:04.341: INFO: stderr: "warning: deleting cluster-scoped resources, not scoped to the provided namespace\n"
Nov 14 11:37:04.341: INFO: stdout: "storageclass.storage.k8s.io \"csi-rbd-sc\" deleted\n"
Nov 14 11:37:04.345: INFO: ExecWithOptions {Command:[/bin/sh -c ceph fsid] Namespace:rook-ceph PodName:rook-ceph-tools-7467d8bf8-62zvp ContainerName:rook-ceph-tools Stdin:<nil> CaptureStdout:true CaptureStderr:true PreserveWhitespace:true Quiet:false}
Nov 14 11:37:04.345: INFO: >>> kubeConfig: /root/.kube/config
Nov 14 11:37:05.259: INFO: Waiting up to &PersistentVolumeClaim{ObjectMeta:{rbd-pvc  rbd-1318    0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] []  []},Spec:PersistentVolumeClaimSpec{AccessModes:[ReadWriteOnce],Resources:ResourceRequirements{Limits:ResourceList{},Requests:ResourceList{storage: {{1073741824 0} {<nil>} 1Gi BinarySI},},},VolumeName:,Selector:nil,StorageClassName:*csi-rbd-sc,VolumeMode:nil,DataSource:nil,DataSourceRef:nil,},Status:PersistentVolumeClaimStatus{Phase:,AccessModes:[],Capacity:ResourceList{},Conditions:[]PersistentVolumeClaimCondition{},},} to be in Bound state
Nov 14 11:37:05.259: INFO: waiting for PVC rbd-pvc (0 seconds elapsed)
Nov 14 11:37:07.262: INFO: waiting for PVC rbd-pvc (2 seconds elapsed)
Nov 14 11:37:07.267: INFO: Waiting for PV pvc-a2efc7e6-68f3-422c-99f2-8bb2ac2eb070 to bind to PVC rbd-pvc
Nov 14 11:37:07.267: INFO: Waiting up to timeout=10m0s for PersistentVolumeClaims [rbd-pvc] to have phase Bound
Nov 14 11:37:07.269: INFO: PersistentVolumeClaim rbd-pvc found and phase=Bound (1.862348ms)
Nov 14 11:37:07.269: INFO: Waiting up to 10m0s for PersistentVolume pvc-a2efc7e6-68f3-422c-99f2-8bb2ac2eb070 to have phase Bound
Nov 14 11:37:07.270: INFO: PersistentVolume pvc-a2efc7e6-68f3-422c-99f2-8bb2ac2eb070 found and phase=Bound (1.769751ms)
Nov 14 11:37:07.278: INFO: Waiting up to csi-rbd-demo-pod to be in Running state
Nov 14 11:37:25.290: INFO: ExecWithOptions {Command:[/bin/sh -c rbd image-meta get replicapool/csi-vol-313c62e6-453f-11ec-81bd-12cf8b4b9b0f rbd.csi.ceph.com/encrypted] Namespace:rook-ceph PodName:rook-ceph-tools-7467d8bf8-62zvp ContainerName:rook-ceph-tools Stdin:<nil> CaptureStdout:true CaptureStderr:true PreserveWhitespace:true Quiet:false}
Nov 14 11:37:25.290: INFO: >>> kubeConfig: /root/.kube/config
Nov 14 11:37:25.687: INFO: ExecWithOptions {Command:[/bin/sh -c lsblk -o TYPE,MOUNTPOINT | grep '/var/lib/www/html' | awk '{print $1}'] Namespace:rbd-1318 PodName:csi-rbd-demo-pod ContainerName:web-server Stdin:<nil> CaptureStdout:true CaptureStderr:true PreserveWhitespace:true Quiet:false}
Nov 14 11:37:25.687: INFO: >>> kubeConfig: /root/.kube/config
Nov 14 11:37:25.816: FAIL: failed to validate encrypted pvc with error  not equal to crypt
```

pkalever commented:
@ceph/ceph-csi-contributors 

I don't know why but the mount type is empty instead of `crypt`, hence it is failing exactly at https://github.com/ceph/ceph-csi/blob/devel/e2e/rbd_helper.go#L556

Have we seen this before? Or has anyone made changes recently in this code path?

Madhu-1 reviewed (commented):
LGTM. as the tests are even failing on krbd commit 3 can be removed?

Madhu-1 commented:
> @pkalever is this test failing manually also if we try 3/4 times?

@Madhu-1 I have tried it locally and checked manually, the pod comes up clean with NBD + encryption with this changes (unlike before)

pkalever commented:
> LGTM. as the tests are even failing on krbd commit 3 can be removed?

Yes, certainly!

pkalever commented:
Before removing commit 3, which is for reordering tests such a way that the `krbd + encryption` tests run before `rbd-nbd + encryption` tests, Just for the record sake, capturing the results links here which failed with `krbd + encryption` ON.

https://jenkins-ceph-csi.apps.ocp.ci.centos.org/job/mini-e2e-helm_k8s-1.20/3079/display/redirect
https://jenkins-ceph-csi.apps.ocp.ci.centos.org/job/mini-e2e-helm_k8s-1.21/1738/display/redirect
https://jenkins-ceph-csi.apps.ocp.ci.centos.org/job/mini-e2e-helm_k8s-1.22/1028/display/redirect
https://jenkins-ceph-csi.apps.ocp.ci.centos.org/job/mini-e2e_k8s-1.20/3097/display/redirect
https://jenkins-ceph-csi.apps.ocp.ci.centos.org/job/mini-e2e_k8s-1.21/1754/display/redirect
https://jenkins-ceph-csi.apps.ocp.ci.centos.org/job/mini-e2e_k8s-1.22/1005/display/redirect

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.22

```
X Exiting due to GUEST_PROVISION: Failed to start host: creating host: create: Error creating machine: Error in driver during machine creation: IP not available after waiting: machine minikube didn't return IP after 1 minute

```

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.22

nixpanic commented:
> Nov 14 11:37:25.687: INFO: ExecWithOptions {Command:[/bin/sh -c lsblk -o TYPE,MOUNTPOINT | grep '/var/lib/www/html' | awk '{print $1}'] Namespace:rbd-1318 PodName:csi-rbd-demo-pod ContainerName:web-server Stdin:<nil> CaptureStdout:true CaptureStderr:true PreserveWhitespace:true Quiet:false}
> Nov 14 11:37:25.687: INFO: >>> kubeConfig: /root/.kube/config
> Nov 14 11:37:25.816: FAIL: failed to validate encrypted pvc with error  not equal to crypt

This test should run on the nodeplugin, not inside the application pod... The application may not see, or have sufficient permissions for detecting the block-device type (encrypted or not).

Commit:
e2e: validate encrypted image mount inside the nodeplugin

currently the mountType validation of the encrypted volume is done in
the application, we should rather validate this inside the nodeplugin
pod.

Signed-off-by: Prasanna Kumar Kalever <prasanna.kalever@redhat.com>

pkalever commented:
> This test should run on the nodeplugin, not inside the application pod... The application may not see, or have sufficient permissions for detecting the block-device type (encrypted or not).

@nixpanic 
Sure, refreshed with the required changes. Thanks!

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.20

```
X Exiting due to GUEST_PROVISION: Failed to start host: creating host: create: Error creating machine: Error in driver during machine creation: IP not available after waiting: machine minikube didn't return IP after 1 minute

```

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.20

nixpanic commented on e2e/rbd_helper.go:
```diff
@@ -548,7 +548,11 @@ func validateEncryptedImage(f *framework.Framework, rbdImageSpec string, app *v1
-	mountType, err := getMountType(app.Name, app.Namespace, volumeMountPath, f)
+	selector, err := getDaemonSetLabelSelector(f, cephCSINamespace, rbdDaemonsetName)
+	if err != nil {
+		e2elog.Failf("failed to get the labels with error %v", err)
```
use `"failed to get labels: %v"`, no need to add `with error` in an error log.

Rakshith-R commented on e2e/rbd_helper.go:
```diff
@@ -547,8 +547,15 @@ func validateEncryptedImage(f *framework.Framework, rbdImageSpec string, app *v1
-	mountType, err := getMountType(app.Name, app.Namespace, volumeMountPath, f)
+	volumeMountPath := fmt.Sprintf(
+		"/var/lib/kubelet/pods/%s/volumes/kubernetes.io~csi/%s/mount",
+		app.UID,
```
This seems to be empty from ci,
App needs to be fetched to update uid.

pkalever commented:
/retest all

Rakshith-R commented on internal/rbd/nodeserver.go:
```diff
@@ -226,14 +226,14 @@ func populateRbdVol(
 		rv.Mounter = req.GetVolumeContext()["mounter"]
 	} else {
 		// fallback to rbd-nbd,
 		// ignore the mapOptions and unmapOptions as they are meant for krbd use.
```
`// ignore the mapOptions and unmapOptions as they are meant for krbd use.`
Please update/correct this comment too.

pkalever replied:
Not really! I will refresh the series in such a way that the comments make sense.

Rakshith-R reviewed (changes requested):
Ci is finally passing 🎊.

Madhu-1 commented on internal/rbd/nodeserver.go:
```diff
@@ -216,23 +216,21 @@ func populateRbdVol(
-	}
-	if krbdSupported {
+	} else {
+		rv.Mounter = req.GetVolumeContext()["mounter"]
```
what if the user has specified generic mount options which are applicable for both krbd/rbd-nbd?

pkalever replied:
Most of the time it would be specific to the mounter type. 

> what if the user has specified generic mount options which are applicable for both krbd/rbd-nbd?

This is why I recommended having mounter-specific mapOptions in the storage class. I don't think we have a way to detect which options belong to which mounter etc, so that we can have one `mapOption`, and adding such a mechanism might lead to additional maintenance hickups.

Madhu-1 replied:
okay. we need to identify a way to achieve this one possible by avoiding adding more SC parameters.

pkalever replied:
Yes! let's take that in a follow-up PR to avoid additional delay with this PR, considering this PR is on the hot seat now :-)

humblec replied:
@pkalever  @Madhu-1  one idea to avoid having more sc parameters are say you specify mounter in the sc, at that time itself specify the mount options 

Like 

`mounter: "nbd:<nbdmapoptions>"` 

or 

`mounter: "krbd:<krbdmapoptions>"`

pkalever replied:
Good Idea! Humble. Let's see what others got.

Madhu-1 replied:
AFAIK this doesn't solve the problem for the tryOthermonters. IMO this can be solved by adding  `mapOption:"kbrd:v1,v2,v2,nbd:v1,v2,v3"` same applies for unmap also?

pkalever replied:
Yes! So its the user's responsibility to tweak mapOptions if he is setting `tryOthermonters: true` in a way that it works for krbd, nbd or some other mounter in the future.

`mapOption: "common: op1, op2 :: krbd-specific: op1, op2 :: nbd-specific: op1, op2"`

delimiters are fancy, can be changed to anything else as per everyone's agreement.

Madhu-1 replied:
Yes, the main problem is to have a proper delimiter for identification

humblec replied:
yeah, individual mounters can take `mapoptions` as mentioned above https://github.com/ceph/ceph-csi/pull/2636#discussion_r749961462, for tryOtherMounters , it can be clubbed as in https://github.com/ceph/ceph-csi/pull/2636#discussion_r749971324.. Which gives enough freedom for the users to have mapoptions specified based on the mounter even in absense of optional settings like `tryOtherMounters`.. :+1:

nixpanic replied:
Its horrible, very user unfriendly... but on the other hand, I also do not like a need for mounter-specific configuration options.

So, I'd like to go with:

```yaml
mapOption: "kbrd:v1,v2,v3;nbd:v1,v2,v3"
```

- By omitting `krbd:` or `nbd:`, the option(s) apply to both.
- A user can _override_ the options for a mounter by specifying `krbd:` or `nbd:`. 
  ```yaml
  mapOption: "v1,v2,v3;nbd:v1,v2,v3"
  ```
  is effectively the same as the 1st example.
- Sections are split by `;`.
- No need to specify `common`, keep things as simple as possible. If users want to specify options for both `krbd` and `nbd`, they can mention them twice them.

humblec replied:
`mapOption` can be kept for common mapping options between all mounters we support, regardless,  I am fine with any of the above models, the main point was that, `using mapping options followed with a delimter to reduce sc params`..

pkalever replied:
@nixpanic well summarized :+1:

pkalever replied:
> By omitting krbd: or nbd:, the option(s) apply to both.

@nixpanic won't this be backward-incompatible ? IMHO this should be treated as options to `rbdDefaultMounter` like before.

pkalever replied:
Here is the issue to track this item https://github.com/ceph/ceph-csi/issues/2641
Please use the issue for any further discussions on this topic.

Madhu-1 commented:
@Rakshith-R PTAL as you have requested for the changes.

Rakshith-R reviewed (dismissed):
Lgtm, 
We can decide on Un/Map Options in follow up pr.

humblec commented on e2e/rbd_helper.go:
```diff
@@ -547,8 +547,21 @@ func validateEncryptedImage(f *framework.Framework, rbdImageSpec string, app *v1
-	mountType, err := getMountType(app.Name, app.Namespace, volumeMountPath, f)
+	pod, err := f.ClientSet.CoreV1().Pods(f.UniqueName).Get(context.TODO(), appName, metav1.GetOptions{})
+	if err != nil {
+		e2elog.Logf("error getting pod %q in namespace %q", appName, f.UniqueName)
```
@pkalever do we need this here? I doubt thats the case.

pkalever replied:
@humblec you mean just delete the logline, right?

pkalever replied:
Done now!

humblec replied:
yep. @pkalever

pkalever replied:
Thanks!

Madhu-1 commented on e2e/rbd_helper.go:
```diff
@@ -547,8 +547,21 @@ func validateEncryptedImage(f *framework.Framework, rbdImageSpec string, app *v1
+		pvName)
+	selector, err := getDaemonSetLabelSelector(f, cephCSINamespace, rbdDaemonsetName)
+	if err != nil {
+		e2elog.Failf("failed to get labels: %v", err)
```
return error instead of failing the test

pkalever commented:
/retest ci/centos/k8s-e2e-external-storage/1.22
/retest ci/centos/mini-e2e-helm/k8s-1.22

```
X Exiting due to GUEST_PROVISION: Failed to start host: creating host: create: Error creating machine: Error in driver during machine creation: IP not available after waiting: machine minikube didn't return IP after 1 minute

```

pkalever commented:
/retest ci/centos/k8s-e2e-external-storage/1.22

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22
